src/D2L.CodeStyle.Analyzers/ApiUsage/Serialization/ReflectionSerializerAnalyzer.cs

```diff
@@ -0,0 +1,318 @@
+
+				context.ReportDiagnostic( diagnostic );
+			}
+		}
+
+		private static void ReportConstructorParameterCannotBeDeserialized(
+				SymbolAnalysisContext context,
+				IParameterSymbol parameter
+			) {
+
+			ParameterSyntax declaration = GetFirstDeclaringSyntax<ParameterSyntax>( context, parameter );
+
+			Diagnostic diagnostic = Diagnostic.Create(
+					descriptor: ReflectionSerializer_ConstructorParameterCannotBeDeserialized,
+					location: declaration.Identifier.GetLocation(),
+					messageArgs: new[] { parameter.Name }
+				);
+
+			context.ReportDiagnostic( diagnostic );
+		}
+
+		private static void AnalyzeRecordProperties(
+				SymbolAnalysisContext context,
+				INamedTypeSymbol type
+			) {
+
+			IEnumerable<IPropertySymbol> instanceProperties = type
+				.GetMembers()
+				.Where( m => !m.IsStatic )
+				.OfType<IPropertySymbol>();
+
+			foreach( IPropertySymbol property in instanceProperties ) {
+				AnalyzeRecordProperty( context, property );
+			}
+		}
+
+		private static void AnalyzeRecordProperty(
+				SymbolAnalysisContext context,
+				IPropertySymbol property
+			) {
+
+			if( property.DeclaringSyntaxReferences.IsEmpty ) {
+				return;
+			}
+
+			SyntaxNode declaringSyntax = property
+				.DeclaringSyntaxReferences[0]
+				.GetSyntax( context.CancellationToken );
+
+			if( !( declaringSyntax is PropertyDeclarationSyntax declaration ) ) {
+				return;
+			}
+
+			AccessorListSyntax accessors = declaration.AccessorList;
+			if( accessors == null ) {
+				return;
+			}
+
+			for( int i = 0; i < accessors.Accessors.Count; i++ ) {
+
+				AccessorDeclarationSyntax accessor = accessors.Accessors[i];
+				if( accessor.Kind() == SyntaxKind.InitAccessorDeclaration ) {
+
+					Diagnostic d = Diagnostic.Create(
+							ReflectionSerializer_Record_InitOnlySetter,
+							accessor.Keyword.GetLocation()
+						);
+
+					context.ReportDiagnostic( d );
+					return;
+				}
+			}
+		}
+
+		private enum TypeDeclarationKind {
+			Class,
+			Record
+		}
+
+		private sealed record TypeDeclaration(
+				TypeDeclarationKind Kind,
+				TypeDeclarationSyntax Syntax
+			);
+
+		private static TypeDeclaration GetTypeDeclaration(
+				SymbolAnalysisContext context,
+				AttributeData reflectionSerializerAttribute
+			) {
+
+			SyntaxNode attribute = reflectionSerializerAttribute
+				.ApplicationSyntaxReference
+				.GetSyntax( context.CancellationToken );
+
+			if( !( attribute.Parent is AttributeListSyntax attributeList ) ) {
+				throw new InvalidOperationException( $"Unexpected parent kind of AttributeSyntax: { attribute.Parent.Kind() }" );
+			}
+
+			switch( attributeList.Parent ) {
+
+				case ClassDeclarationSyntax @class:
+					return new( TypeDeclarationKind.Class, @class );
+
+				case RecordDeclarationSyntax record:
+					return new( TypeDeclarationKind.Record, record );
+
+				default:
+					throw new NotSupportedException( $"Unsupported [ReflectionSerializer] attribute target kind: { attributeList.Parent.Kind() }" );
+			}
+		}

```

Comment:
Newer CodeAnalysis has `INamedTypeSymbol.IsRecord`, but we're not updated. It has to sync with Visual Studio requirements so can't just do so willy-nilly, but this is something to replace when we do update.

Reply:
Going to try rolling out an update: https://github.com/Brightspace/D2L.CodeStyle/pull/769

Reply:
Updated in https://github.com/Brightspace/D2L.CodeStyle/pull/771 which is successfully building in CI now.

Consider simplifying this with `INamedTypeSymbol.IsRecord`, and then just putting the diagnostic on the AttributeSyntax instead? The less we're traversing syntax, the less edge cases we have (analyzers crashing or throwing exceptions have a very poor experience).

Reply:
Arg, eff, nevermind. Version update got finicky later on in the build.

Carry on.